File: src/cli/freePlan.ts
/**
 * ZEESH FREE session display helpers (Milestone 13).
 *
 * Pure rendering only — every number comes from the backend's daily session
 * state (GET /api/usage / POST /api/provider response), so the CLI never
 * trusts or stores quota locally. All functions degrade gracefully: pass null
 * / invalid data and they return nothing to print.
 */
import type { DailySessionState } from '../auth/client.ts';
import { c } from './colors.ts';

/** Render seconds as a compact countdown, e.g. 47m 12s ('' when invalid). */
export function formatCountdown(seconds: number | null | undefined): string {
  if (typeof seconds !== 'number' || !Number.isFinite(seconds)) return '';
  const total = Math.max(0, Math.floor(seconds));
  if (total <= 0) return 'expired';
  const m = Math.floor(total / 60);
  const s = total % 60;
  return m > 0 ? `${m}m ${s}s` : `${s}s`;
}

/**
 * Render daily usage totals with hours, e.g. 6h, 1h 35m, 10m, 45s
 * ('' when invalid). formatDuration only goes up to minutes — daily
 * totals need the hour unit.
 */
export function formatDailyUsage(seconds: number | null | undefined): string {
  if (typeof seconds !== 'number' || !Number.isFinite(seconds) || seconds < 0) return '';
  const total = Math.floor(seconds);
  const h = Math.floor(total / 3600);
  const m = Math.floor((total % 3600) / 60);
  const s = total % 60;
  if (h > 0) return m > 0 ? `${h}h ${m}m` : `${h}h`;
  if (m > 0) return `${m}m`;
  return `${s}s`;
}

/**
 * Seconds left in the current session from an ISO expiresAt timestamp
 * (computed locally for display only — enforcement stays server-side).
 */
export function sessionSecondsLeft(expiresAt: string | null | undefined): number | null {
  if (!expiresAt) return null;
  const t = new Date(expiresAt).getTime();
  if (Number.isNaN(t)) return null;
  return Math.max(0, Math.round((t - Date.now()) / 1000));
}

/**
 * The "Session X / 6 · time remaining · today's usage" line shown after each
 * run. Returns '' when there is no session state to display.
 */
export function sessionStatusLine(
  state: DailySessionState | null | undefined,
): string {
  if (!state || typeof state.sessionsUsed !== 'number') return '';
  const total = state.sessionsUsed + state.sessionsRemaining;
  const left = sessionSecondsLeft(state.sessionExpiresAt);
  const remaining = left !== null ? formatCountdown(left) : 'no active session';
  const usedToday = formatDailyUsage(state.dailyUsedSeconds);
  const limitToday = formatDailyUsage(state.dailyLimitSeconds);
  return c.dim(`Session ${state.currentSession ?? state.sessionsUsed} / ${total} · ${remaining} left · ${usedToday} / ${limitToday} used today`);
}

/** Note shown when the server rolled the user into a fresh session. */
export function sessionRolloverNote(state: DailySessionState | null | undefined): string {
  if (!state) return '';
  return c.dim(`Session ${state.currentSession} of ${state.sessionsUsed + state.sessionsRemaining} started — the previous session expired.`);
}

/**
 * Short banner row: "Free plan: 2/6 sessions · 41m left" or a simpler
 * availability line when no session is active yet. '' when unavailable.
 */
export function bannerFreePlanLine(state: DailySessionState | null | undefined): string {
  if (!state || typeof state.sessionsUsed !== 'number') return '';
  const total = state.sessionsUsed + state.sessionsRemaining;
  if (state.currentSession != null && state.sessionExpiresAt) {
    const left = formatCountdown(sessionSecondsLeft(state.sessionExpiresAt));
    return c.green(`Free plan · Session ${state.currentSession}/${total} · ${left} left · ${formatDailyUsage(state.dailyUsedSeconds)} used today`);
  }
  if (state.sessionsRemaining === 0) {
    return c.yellow(`Free plan · all ${total} sessions used today — more at 00:00 UTC`);
  }
  return c.green(`Free plan · ${state.sessionsRemaining} of ${total} sessions available today (${formatDailyUsage(state.dailyLimitSeconds)} max)`);
}

File: grace/cli/free_plan.py
"""GRACE FREE session display helpers (port of src/cli/freePlan.ts).

Pure rendering only — every number comes from the backend's daily session
state (GET /api/usage / POST /api/provider response), so the CLI never trusts
or stores quota locally. All functions degrade gracefully: pass None / invalid
data and they return nothing to print.
"""

import math
from datetime import datetime, timezone

from grace.colors import c


def format_countdown(seconds) -> str:
    """Render seconds as a compact countdown, e.g. 47m 12s ('' when invalid)."""
    if not isinstance(seconds, (int, float)) or seconds != seconds:  # NaN check
        return ""
    total = max(0, math.floor(seconds))
    if total <= 0:
        return "expired"
    m = total // 60
    s = total % 60
    return f"{m}m {s}s" if m > 0 else f"{s}s"


def format_daily_usage(seconds) -> str:
    """Render daily usage totals with hours, e.g. 6h, 1h 35m, 10m, 45s."""
    if not isinstance(seconds, (int, float)) or seconds != seconds or seconds < 0:
        return ""
    total = math.floor(seconds)
    h = total // 3600
    m = (total % 3600) // 60
    s = total % 60
    if h > 0:
        return f"{h}h {m}m" if m > 0 else f"{h}h"
    if m > 0:
        return f"{m}m"
    return f"{s}s"


def session_seconds_left(expires_at) -> int | None:
    """Seconds left in the current session from an ISO expiresAt timestamp
    (computed locally for display only — enforcement stays server-side)."""
    if not expires_at:
        return None
    try:
        t = datetime.fromisoformat(str(expires_at).replace("Z", "+00:00")).timestamp() * 1000
    except Exception:
        return None
    return max(0, round((t - datetime.now(timezone.utc).timestamp() * 1000) / 1000))


def session_status_line(state) -> str:
    """The \"Session X / 6 · time remaining · today's usage\" line shown after
    each run. Returns '' when there is no session state to display."""
    if not state or not isinstance(state.get("sessionsUsed"), (int, float)):
        return ""
    total = state["sessionsUsed"] + (state.get("sessionsRemaining") or 0)
    left = session_seconds_left(state.get("sessionExpiresAt"))
    remaining = format_countdown(left) if left is not None else "no active session"
    used_today = format_daily_usage(state.get("dailyUsedSeconds"))
    limit_today = format_daily_usage(state.get("dailyLimitSeconds"))
    return c.dim(f"Session {state.get('currentSession') or state['sessionsUsed']} / {total} · {remaining} left · {used_today} / {limit_today} used today")


def session_rollover_note(state) -> str:
    """Note shown when the server rolled the user into a fresh session."""
    if not state:
        return ""
    return c.dim(f"Session {state.get('currentSession')} of {state.get('sessionsUsed', 0) + (state.get('sessionsRemaining') or 0)} started — the previous session expired.")


def banner_free_plan_line(state) -> str:
    """Short banner row: \"Quota: 5 sessions remaining · 58m left\" or a simpler
    availability line when no session is active yet. '' when unavailable."""
    if not state or not isinstance(state.get("sessionsUsed"), (int, float)):
        return ""
    total = state["sessionsUsed"] + (state.get("sessionsRemaining") or 0)
    if state.get("currentSession") is not None and state.get("sessionExpiresAt"):
        left = format_countdown(session_seconds_left(state["sessionExpiresAt"]))
        return c.green(f"Quota · Session {state['currentSession']}/{total} · {left} left · {format_daily_usage(state.get('dailyUsedSeconds'))} used today")
    if (state.get("sessionsRemaining") or 0) == 0:
        return c.yellow(f"Quota · all {total} sessions used today — more at 00:00 UTC")
    return c.green(f"Quota · {state.get('sessionsRemaining')} sessions remaining today ({format_daily_usage(state.get('dailyLimitSeconds'))} max)")
